--- a/Python/dynload_win.c
+++ b/Python/dynload_win.c
@@ -107,7 +107,7 @@ static char *GetPythonImport (HINSTANCE hModule)
                 char *pch;
 
                 /* Don't claim that python3.dll is a Python DLL. */
-#ifdef _DEBUG
+#if 0
                 if (strcmp(import_name, "python3_d.dll") == 0) {
 #else
                 if (strcmp(import_name, "python3.dll") == 0) {
@@ -119,7 +119,7 @@ static char *GetPythonImport (HINSTANCE hModule)
                 /* Ensure python prefix is followed only
                    by numbers to the end of the basename */
                 pch = import_name + 6;
-#ifdef _DEBUG
+#if 0
                 while (*pch && pch[0] != '_' && pch[1] != 'd' && pch[2] != '.') {
 #else
                 while (*pch && *pch != '.') {
@@ -158,12 +158,16 @@ _Py_CheckPython3(void)
     static int python3_checked = 0;
     static HANDLE hPython3;
     #define MAXPATHLEN 512
-    wchar_t py3path[MAXPATHLEN+1];
     if (python3_checked) {
         return hPython3 != NULL;
     }
     python3_checked = 1;
 
+#ifdef MS_WINDOWS_APP
+    // LoadPackagedLibrary doesn't accept absolute paths so load dll name from current app dir
+    hPython3 = LoadPackagedLibrary(PY3_DLLNAME, 0);
+#else
+    wchar_t py3path[MAXPATHLEN + 1];
     /* If there is a python3.dll next to the python3y.dll,
        use that DLL */
     if (PyWin_DLLhModule && GetModuleFileNameW(PyWin_DLLhModule, py3path, MAXPATHLEN)) {
@@ -195,6 +199,8 @@ _Py_CheckPython3(void)
             hPython3 = LoadLibraryExW(py3path, NULL, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS);
         }
     }
+#endif
+
     return hPython3 != NULL;
     #undef MAXPATHLEN
 }
@@ -231,9 +237,19 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
            AddDllDirectory function. We add SEARCH_DLL_LOAD_DIR to
            ensure DLLs adjacent to the PYD are preferred. */
         Py_BEGIN_ALLOW_THREADS
+#ifdef MS_WINDOWS_APP
+        // LoadPackagedLibrary doesn't accept absolute paths so use
+        // the base path or DLL name
+        wchar_t* base_path = wcsstr(wpathname, L"system");
+        if (base_path)
+            hDLL = LoadPackagedLibrary(base_path, 0);
+        else
+            hDLL = LoadPackagedLibrary(wpathname, 0);
+#else
         hDLL = LoadLibraryExW(wpathname, NULL,
                               LOAD_LIBRARY_SEARCH_DEFAULT_DIRS |
                               LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR);
+#endif
         Py_END_ALLOW_THREADS
         PyMem_Free(wpathname);
 
@@ -299,7 +315,7 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
             char buffer[256];
 
             PyOS_snprintf(buffer, sizeof(buffer),
-#ifdef _DEBUG
+#if 0
                           "python%d%d_d.dll",
 #else
                           "python%d%d.dll",
--- a/Python/fileutils.c
+++ b/Python/fileutils.c
@@ -1100,50 +1100,40 @@ typedef union {
 
 
 void
-_Py_attribute_data_to_stat(BY_HANDLE_FILE_INFORMATION *info, ULONG reparse_tag,
-                           FILE_BASIC_INFO *basic_info, FILE_ID_INFO *id_info,
-                           struct _Py_stat_struct *result)
+_Py_attribute_data_to_stat(FILE_BASIC_INFO* basic_info,
+    FILE_STANDARD_INFO* standard_info,
+    ULONG reparse_tag,
+    struct _Py_stat_struct* result)
 {
     memset(result, 0, sizeof(*result));
-    result->st_mode = attributes_to_mode(info->dwFileAttributes);
-    result->st_size = (((__int64)info->nFileSizeHigh)<<32) + info->nFileSizeLow;
-    result->st_dev = id_info ? id_info->VolumeSerialNumber : info->dwVolumeSerialNumber;
+    result->st_mode = attributes_to_mode(basic_info->FileAttributes);
+    result->st_size = standard_info->EndOfFile.QuadPart;
+    result->st_dev = 1;
     result->st_rdev = 0;
+
     /* st_ctime is deprecated, but we preserve the legacy value in our caller, not here */
-    if (basic_info) {
-        LARGE_INTEGER_to_time_t_nsec(&basic_info->CreationTime, &result->st_birthtime, &result->st_birthtime_nsec);
-        LARGE_INTEGER_to_time_t_nsec(&basic_info->ChangeTime, &result->st_ctime, &result->st_ctime_nsec);
-        LARGE_INTEGER_to_time_t_nsec(&basic_info->LastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
-        LARGE_INTEGER_to_time_t_nsec(&basic_info->LastAccessTime, &result->st_atime, &result->st_atime_nsec);
-    } else {
-        FILE_TIME_to_time_t_nsec(&info->ftCreationTime, &result->st_birthtime, &result->st_birthtime_nsec);
-        FILE_TIME_to_time_t_nsec(&info->ftLastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
-        FILE_TIME_to_time_t_nsec(&info->ftLastAccessTime, &result->st_atime, &result->st_atime_nsec);
-    }
-    result->st_nlink = info->nNumberOfLinks;
+    LARGE_INTEGER_to_time_t_nsec(&basic_info->CreationTime, &result->st_birthtime, &result->st_birthtime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&basic_info->ChangeTime, &result->st_ctime, &result->st_ctime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&basic_info->LastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&basic_info->LastAccessTime, &result->st_atime, &result->st_atime_nsec);
+    result->st_nlink = standard_info->NumberOfLinks;
 
-    if (id_info) {
-        id_128_to_ino file_id;
-        file_id.id = id_info->FileId;
-        result->st_ino = file_id.st_ino;
-        result->st_ino_high = file_id.st_ino_high;
-    }
     if (!result->st_ino && !result->st_ino_high) {
         /* should only occur for DirEntry_from_find_data, in which case the
            index is likely to be zero anyway. */
-        result->st_ino = (((uint64_t)info->nFileIndexHigh) << 32) + info->nFileIndexLow;
+        result->st_ino = basic_info->CreationTime.QuadPart;
     }
 
     /* bpo-37834: Only actual symlinks set the S_IFLNK flag. But lstat() will
        open other name surrogate reparse points without traversing them. To
        detect/handle these, check st_file_attributes and st_reparse_tag. */
     result->st_reparse_tag = reparse_tag;
-    if (info->dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+    if (basic_info->FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
         reparse_tag == IO_REPARSE_TAG_SYMLINK) {
         /* set the bits that make this a symlink */
         result->st_mode = (result->st_mode & ~S_IFMT) | S_IFLNK;
     }
-    result->st_file_attributes = info->dwFileAttributes;
+    result->st_file_attributes = basic_info->FileAttributes;
 }
 
 void
@@ -1210,6 +1200,61 @@ _Py_stat_basic_info_to_stat(FILE_STAT_BASIC_INFORMATION *info,
     }
 }
 
+void
+_Py_find_data_to_stat(WIN32_FIND_DATAW* find_data,
+    struct _Py_stat_struct* result)
+{
+    memset(result, 0, sizeof(*result));
+    result->st_mode = attributes_to_mode(find_data->dwFileAttributes);
+    FILE_TIME_to_time_t_nsec(&find_data->ftCreationTime, &result->st_ctime, &result->st_ctime_nsec);
+    FILE_TIME_to_time_t_nsec(&find_data->ftLastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
+    FILE_TIME_to_time_t_nsec(&find_data->ftLastAccessTime, &result->st_atime, &result->st_atime_nsec);
+    result->st_size = ((long long)find_data->nFileSizeHigh) << 32 > find_data->nFileSizeLow;
+    result->st_dev = 0;
+    result->st_rdev = result->st_dev;
+    result->st_nlink = 0;
+    result->st_ino = 0;
+    if (find_data->dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+        find_data->dwReserved0 == IO_REPARSE_TAG_SYMLINK) {
+        /* first clear the S_IFMT bits */
+        result->st_mode ^= (result->st_mode & S_IFMT);
+        /* now set the bits that make this a symlink */
+        result->st_mode |= S_IFLNK;
+    }
+    result->st_file_attributes = find_data->dwFileAttributes;
+}
+
+int
+_Py_stat_from_file_handle(HANDLE h, struct _Py_stat_struct* result, BOOL set_ino)
+{
+    FILE_BASIC_INFO basic_info = { 0 };
+    FILE_STANDARD_INFO standard_info = { 0 };
+    if (!GetFileInformationByHandleEx(h, FileBasicInfo, &basic_info, sizeof(basic_info))
+        || !GetFileInformationByHandleEx(h, FileStandardInfo, &standard_info, sizeof(standard_info))) {
+        /* The Win32 error is already set, but we also set errno for
+           callers who expect it */
+        switch (GetLastError()) {
+        case ERROR_INVALID_PARAMETER:
+        case ERROR_INVALID_FUNCTION:
+        case ERROR_NOT_SUPPORTED:
+            /* Volumes and physical disks are block devices, e.g.
+               \\.\C: and \\.\PhysicalDrive0. */
+            memset(result, 0, sizeof(*result));
+            result->st_mode = 0x6000; /* S_IFBLK */
+        }
+        PyErr_SetFromWindowsErr(0);
+        errno = winerror_to_errno(GetLastError());
+        return -1;
+    }
+
+    _Py_attribute_data_to_stat(&basic_info, &standard_info, 0, result);
+    /* specific to fstat() */
+    if (set_ino) {
+        result->st_ino = basic_info.CreationTime.QuadPart;
+    }
+    return 0;
+}
+
 #endif
 
 /* Return information about a file.
@@ -1228,8 +1273,6 @@ int
 _Py_fstat_noraise(int fd, struct _Py_stat_struct *status)
 {
 #ifdef MS_WINDOWS
-    BY_HANDLE_FILE_INFORMATION info;
-    FILE_BASIC_INFO basicInfo;
     FILE_ID_INFO idInfo;
     FILE_ID_INFO *pIdInfo = &idInfo;
     HANDLE h;
@@ -1263,21 +1306,12 @@ _Py_fstat_noraise(int fd, struct _Py_stat_struct *status)
         return 0;
     }
 
-    if (!GetFileInformationByHandle(h, &info) ||
-        !GetFileInformationByHandleEx(h, FileBasicInfo, &basicInfo, sizeof(basicInfo))) {
-        /* The Win32 error is already set, but we also set errno for
-           callers who expect it */
-        errno = winerror_to_errno(GetLastError());
-        return -1;
-    }
-
     if (!GetFileInformationByHandleEx(h, FileIdInfo, &idInfo, sizeof(idInfo))) {
         /* Failed to get FileIdInfo, so do not pass it along */
         pIdInfo = NULL;
     }
 
-    _Py_attribute_data_to_stat(&info, 0, &basicInfo, pIdInfo, status);
-    return 0;
+    return _Py_stat_from_file_handle(h, status, TRUE);
 #else
     return fstat(fd, status);
 #endif
@@ -1398,6 +1432,9 @@ static int
 get_inheritable(int fd, int raise)
 {
 #ifdef MS_WINDOWS
+#ifdef MS_WINDOWS_APP
+    return 0;
+#endif
     HANDLE handle;
     DWORD flags;
 
@@ -1484,11 +1521,13 @@ set_inheritable(int fd, int inheritable, int raise, int *atomic_flag_works)
     else
         flags = 0;
 
+#ifndef MS_WINDOWS_APP
     if (!SetHandleInformation(handle, HANDLE_FLAG_INHERIT, flags)) {
         if (raise)
             PyErr_SetFromWindowsErr(0);
         return -1;
     }
+#endif
     return 0;
 
 #else
--- a/Python/sysmodule.c
+++ b/Python/sysmodule.c
@@ -1735,7 +1735,6 @@ sys_getwindowsversion_impl(PyObject *module)
 {
     PyObject *version;
     int pos = 0;
-    OSVERSIONINFOEXW ver;
 
     if (PyObject_GetOptionalAttrString(module, "_cached_windows_version", &version) < 0) {
         return NULL;
@@ -1745,6 +1744,8 @@ sys_getwindowsversion_impl(PyObject *module)
     }
     Py_XDECREF(version);
 
+    OSVERSIONINFOEXW ver;
+    ZeroMemory(&ver, sizeof(ver));
     ver.dwOSVersionInfoSize = sizeof(ver);
     if (!GetVersionExW((OSVERSIONINFOW*) &ver))
         return PyErr_SetFromWindowsErr(0);
@@ -1772,6 +1773,7 @@ sys_getwindowsversion_impl(PyObject *module)
     SET_VERSION_INFO(PyLong_FromLong(ver.wSuiteMask));
     SET_VERSION_INFO(PyLong_FromLong(ver.wProductType));
 
+#ifndef MS_WINDOWS_APP
     // GetVersion will lie if we are running in a compatibility mode.
     // We need to read the version info from a system file resource
     // to accurately identify the OS version. If we fail for any reason,
@@ -1791,6 +1793,7 @@ sys_getwindowsversion_impl(PyObject *module)
     }
 
     SET_VERSION_INFO(realVersion);
+#endif
 
 #undef SET_VERSION_INFO
 
