--- a/Modules/termios.c
+++ b/Modules/termios.c
@@ -242,6 +242,7 @@
     Py_RETURN_NONE;
 }
 
+#if !defined(__ANDROID__)
 /*[clinic input]
 termios.tcdrain
 
@@ -262,6 +263,7 @@
 
     Py_RETURN_NONE;
 }
+#endif
 
 /*[clinic input]
 termios.tcflush
@@ -320,7 +322,9 @@
     TERMIOS_TCGETATTR_METHODDEF
     TERMIOS_TCSETATTR_METHODDEF
     TERMIOS_TCSENDBREAK_METHODDEF
+#if !defined(__ANDROID__)
     TERMIOS_TCDRAIN_METHODDEF
+#endif
     TERMIOS_TCFLUSH_METHODDEF
     TERMIOS_TCFLOW_METHODDEF
     {NULL, NULL}
