--- a/configure.ac
+++ b/configure.ac
@@ -330,6 +330,9 @@
 	*-apple-ios*)
 		ac_sys_system=iOS
 		;;
+	*-*-darwin*)
+		ac_sys_system=Darwin
+		;;
 	*-*-vxworks*)
 	    ac_sys_system=VxWorks
 	    ;;
@@ -771,6 +774,8 @@
 				;;
 		esac
 		;;
+	*-*-darwin*)
+		;;
 	*-*-vxworks*)
 		_host_ident=$host_cpu
 		;;
@@ -851,6 +856,8 @@
   # On iOS, defining _POSIX_C_SOURCE also disables platform specific features.
   iOS/*)
     define_xopen_source=no;;
+  Darwin/*)
+    define_xopen_source=no;;
   # On QNX 6.3.2, defining _XOPEN_SOURCE prevents netdb.h from
   # defining NI_NUMERICHOST.
   QNX/6.3.2)
