--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -196,8 +196,8 @@
 #else
 /* Unix functions that the configure script doesn't check for */
 #ifndef __VXWORKS__
-#define HAVE_EXECV      1
-#define HAVE_FORK       1
+//#define HAVE_EXECV      0
+//#define HAVE_FORK       0
 #if defined(__USLC__) && defined(__SCO_VERSION__)       /* SCO UDK Compiler */
 #define HAVE_FORK1      1
 #endif
@@ -210,8 +210,8 @@
 #define HAVE_KILL       1
 #define HAVE_OPENDIR    1
 #define HAVE_PIPE       1
-#define HAVE_SYSTEM     1
-#define HAVE_WAIT       1
+//#define HAVE_SYSTEM     0
+//#define HAVE_WAIT       0
 #define HAVE_TTYNAME    1
 #endif  /* _MSC_VER */
 #endif  /* ! __WATCOMC__ || __QNX__ */
