--- a/lib/ldb/wscript
+++ b/lib/ldb/wscript
@@ -519,10 +519,6 @@
                          install=False)
 
         bld.SAMBA_BINARY('ldb_key_value_sub_txn_tdb_test',
-                         bld.SUBDIR('ldb_key_value',
-                             '''ldb_kv_search.c
-                                ldb_kv_index.c
-                                ldb_kv_cache.c''') +
                          'tests/ldb_key_value_sub_txn_test.c',
                          cflags='-DTEST_BE=\"tdb\"',
                          deps='cmocka ldb ldb_tdb_err_map',
--- a/third_party/heimdal_build/wscript_build
+++ b/third_party/heimdal_build/wscript_build
@@ -5,6 +5,7 @@
 from samba_utils import SET_TARGET_TYPE
 from samba_autoconf import CURRENT_CFLAGS
 from samba_utils import LOAD_ENVIRONMENT, TO_LIST
+from samba_bundled import BUILTIN_LIBRARY
 
 def heimdal_path(p, absolute=False):
     hpath = os.path.join("../heimdal", p)
@@ -272,7 +273,10 @@
                                                         extra_cflags=cflags)
     source = heimdal_paths(source)
 
-    obj_target = binname + '.heimdal.objlist'
+    if BUILTIN_LIBRARY(bld, binname):
+        obj_target = binname
+    else:
+        obj_target = binname + '.objlist'
 
     HEIMDAL_SUBSYSTEM(obj_target,
                       source         = source,
