--- a/setup.py
+++ b/setup.py
@@ -252,16 +252,16 @@
     "Crypto.Signature",
     "Crypto.Util",
     "Crypto.Math",
-    "Crypto.SelfTest",
-    "Crypto.SelfTest.Cipher",
-    "Crypto.SelfTest.Hash",
-    "Crypto.SelfTest.IO",
-    "Crypto.SelfTest.Protocol",
-    "Crypto.SelfTest.PublicKey",
-    "Crypto.SelfTest.Random",
-    "Crypto.SelfTest.Signature",
-    "Crypto.SelfTest.Util",
-    "Crypto.SelfTest.Math",
+#    "Crypto.SelfTest",
+#    "Crypto.SelfTest.Cipher",
+#    "Crypto.SelfTest.Hash",
+#    "Crypto.SelfTest.IO",
+#    "Crypto.SelfTest.Protocol",
+#    "Crypto.SelfTest.PublicKey",
+#    "Crypto.SelfTest.Random",
+#    "Crypto.SelfTest.Signature",
+#    "Crypto.SelfTest.Util",
+#    "Crypto.SelfTest.Math",
 ]
 package_data = {
     "Crypto" : [ "py.typed", "*.pyi" ],
