--- a/numpy/distutils/unixccompiler.py
+++ b/numpy/distutils/unixccompiler.py
@@ -36,6 +36,11 @@
         from distutils.sysconfig import get_config_vars
         opt = " ".join(os.environ['OPT'].split())
         gcv_opt = " ".join(get_config_vars('OPT')[0].split())
+        try:
+            self.compiler_so.remove('-mmacosx-version-min=10.15.4')
+            self.compiler_so.remove('-I/Users/Shared/xbmc-depends/x86_64-darwin19.4.0-native/include')
+        except:
+            pass
         ccomp_s = " ".join(self.compiler_so)
         if opt not in ccomp_s:
             ccomp_s = ccomp_s.replace(gcv_opt, opt)
