--- a/src/libcec/cmake/CheckPlatformSupport.cmake
+++ b/src/libcec/cmake/CheckPlatformSupport.cmake
@@ -23,7 +23,7 @@
 #       RPI_LIB_DIR               PATH to Raspberry Pi libs
 #       HAVE_TEGRA_API            ON if Tegra is supported
 #
-
+message(WARNING "CheckPlatformSupport RUNNING")
 set(PLATFORM_LIBREQUIRES "")
 
 include(CheckFunctionExists)
@@ -46,6 +46,7 @@
 
   if("${MSVC_C_ARCHITECTURE_ID}" STREQUAL "X86")
     set(LIB_INFO "${LIB_INFO} (x86)")
+message(WARNING "WIN32: SETTING _USE_32BIT_TIME_T")
     add_definitions(-D_USE_32BIT_TIME_T)
     # force python2 for eventghost
     set(PYTHON_USE_VERSION 2)
--- a/src/libcec/CMakeLists.txt
+++ b/src/libcec/CMakeLists.txt
@@ -164,6 +164,7 @@
   if ((${WIN64}) OR (${_M_ARM64}))
     string(REPLACE "/arch:SSE2" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
   else()
+message(WARNING "WIN32: SETTING _USE_32BIT_TIME_T\nWIN64: ${WIN64}\n_M_ARM64: ${_M_ARM64}")
     add_definitions(-D_USE_32BIT_TIME_T)
   endif()
 else()
