--- a/src/libcec/CMakeLists.txt
+++ b/src/libcec/CMakeLists.txt
@@ -161,9 +161,10 @@
   configure_file(${CMAKE_CURRENT_SOURCE_DIR}/libcec.rc.in ${CMAKE_CURRENT_SOURCE_DIR}/libcec.rc)
   add_definitions(-DDLL_EXPORT)
   add_definitions(-DTARGET_WINDOWS -DNOMINMAX -D_CRT_SECURE_NO_WARNINGS -D_WINSOCKAPI_)
-  if ((${WIN64}) OR (${_M_ARM64}))
+  if ((${WIN64} OR ${AMD64}) OR (${_M_ARM64}))
     string(REPLACE "/arch:SSE2" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
   else()
+message(WARNING "WIN32: SETTING _USE_32BIT_TIME_T\nWIN64: ${WIN64}\n_M_ARM64: ${_M_ARM64}")
     add_definitions(-D_USE_32BIT_TIME_T)
   endif()
 else()
