--- a/src/device.c
+++ b/src/device.c
@@ -123,6 +123,8 @@ int dvdcss_use_ioctls( dvdcss_t dvdcss )
     if( ulMode & OPEN_FLAGS_DASD )
         return 1;
 
+    return 0;
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
     return 0;
 #else
     struct stat fileinfo;
--- a/src/ioctl.c
+++ b/src/ioctl.c
@@ -223,10 +223,11 @@ int ioctl_ReadCopyright( int i_fd, int i_layer, int *pi_copyright )
                         p_buffer, sizeof(p_buffer), &ulDataLen);
 
     *pi_copyright = p_buffer[ 4 ];
-
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+    *pi_copyright = 0;
+    i_ret = 0;
 #else
 #   error "DVD ioctls are unavailable on this system"
-
 #endif
     return i_ret;
 }
@@ -356,7 +357,8 @@ int ioctl_ReadDiscKey( int i_fd, const int *pi_agid, uint8_t *p_key )
     }
 
     memcpy( p_key, p_buffer + 4, DVD_DISCKEY_SIZE );
-
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -487,6 +489,8 @@ int ioctl_ReadTitleKey( int i_fd, const int *pi_agid, int i_pos, uint8_t *p_key
 
     memcpy( p_key, p_buffer + 5, DVD_KEY_SIZE );
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -584,6 +588,8 @@ int ioctl_ReportAgid( int i_fd, int *pi_agid )
 
     *pi_agid = p_buffer[ 7 ] >> 6;
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -692,6 +698,8 @@ int ioctl_ReportChallenge( int i_fd, const int *pi_agid, uint8_t *p_challenge )
 
     memcpy( p_challenge, p_buffer + 4, DVD_CHALLENGE_SIZE );
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -802,6 +810,8 @@ int ioctl_ReportASF( int i_fd, int *pi_asf )
 
     *pi_asf = p_buffer[ 7 ] & 1;
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -905,6 +915,8 @@ int ioctl_ReportKey1( int i_fd, const int *pi_agid, uint8_t *p_key )
 
     memcpy( p_key, p_buffer + 4, DVD_KEY_SIZE );
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -988,6 +1000,9 @@ int ioctl_InvalidateAgid( int i_fd, int *pi_agid )
     i_ret = DosDevIOCtl(i_fd, IOCTL_CDROMDISK, CDROMDISK_EXECMD,
                         &sdc, sizeof(sdc), &ulParamLen,
                         NULL, 0, &ulDataLen);
+
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -1288,6 +1303,8 @@ int ioctl_SendChallenge( int i_fd, const int *pi_agid, const uint8_t *p_challeng
                          &sdc, sizeof(sdc), &ulParamLen,
                          p_buffer, sizeof(p_buffer), &ulDataLen );
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -1397,6 +1414,8 @@ int ioctl_SendKey2( int i_fd, const int *pi_agid, const uint8_t *p_key )
                          &sdc, sizeof(sdc), &ulParamLen,
                          p_buffer, sizeof(p_buffer), &ulDataLen );
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
@@ -1524,6 +1543,8 @@ int ioctl_ReportRPC( int i_fd, int *p_type, int *p_mask, int *p_scheme )
     *p_mask = p_buffer[ 5 ];
     *p_scheme = p_buffer[ 6 ];
 
+#elif defined(__APPLE__) && (defined(__arm__) || defined(__aarch64__))
+// nothing
 #else
 #   error "DVD ioctls are unavailable on this system"
 
